src/Microsoft.AspNet.Mvc.Razor.Host/Directives/CodeTreeCache.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.AspNet.FileProviders;
+using Microsoft.AspNet.Razor.Generator.Compiler;
+using Microsoft.Framework.Expiration.Interfaces;
+
+namespace Microsoft.AspNet.Mvc.Razor.Directives
+{
+    /// <summary>
+    /// Default implementation of <see cref="ICodeTreeCache"/>.
+    /// </summary>
+    public class CodeTreeCache : ICodeTreeCache
+    {
+        private readonly IFileProvider _fileProvider;
+        private readonly ConcurrentDictionary<string, CodeTreeCacheEntry> _codeTreeCache;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CodeTreeCache"/>.
+        /// </summary>
+        /// <param name="fileProvider">The application's <see cref="IFileProvider"/>.</param>
+        public CodeTreeCache(IFileProvider fileProvider)
+        {
+            _fileProvider = fileProvider;
+            _codeTreeCache = new ConcurrentDictionary<string, CodeTreeCacheEntry>(StringComparer.Ordinal);
+        }
+
+        /// <inheritdoc />
+        public CodeTree GetOrAdd([NotNull] string pagePath,
+                                 [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
+        {
+            CodeTreeCacheEntry cacheEntry;
+            if (_codeTreeCache.TryGetValue(pagePath, out cacheEntry))
```

Comment:
For tooling, the cache should get GCed along with the host once they discard the `RazorEditorParser` that hold on to these instances once the page gets closed. For runtime, there's certainly issues of things getting pinned over long term. Alternatives to this might be 

* Implement this interface in Microsoft.AspNet.Mvc.Razor (that already depends on MemoryCache) and use that as the runtime cache type.
* Have it be a pass through cache in runtime - we weren't seeing perf issues with this at runtime and we'd get the behavior we currently have.